### PR TITLE
Add Excon instrumentation

### PR DESCRIPTION
**Pull request by frederikspang**
## Description
Describe your changes:
Adds instrumentation for https://github.com/excon/excon using Excon's Middleware model.

I wanted badly to use with_child_span, however, that seems impossible with the current middleware implementation for Excon.

### CODE REVIEW

File: sentry-ruby/lib/sentry/excon/middleware.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        request_info = extract_request_info(env)
+
+        if propagate_trace?(request_info[:url])
+          set_propagation_headers(env[:headers])
+        end
+      end
+
+      def finish_transaction(response)
+        return if @span.nil?
```
Comment (solnic) [inline sentry-ruby/lib/sentry/excon/middleware.rb:44]:
```suggestion
        return unless @span
```

**Review by solnic — APPROVED** [review]
Thank you! This is looking good.

Screenshot 2024-10-29 at 14 37 00
Screenshot 2024-10-29 at 14 37 52
Screenshot 2024-10-29 at 14 38 29
Screenshot 2024-10-29 at 14 38 51

**Comment by solnic** [comment]
@frederikspang hmm seems like these failures are legit:

```
1) Sentry::Excon with tracing enabled with config.send_default_pii = true records the request's span with query string in data
     Failure/Error:
       expect(request_span.data).to eq({
         "http.response.status_code" => 200,
         "url" => "http://example.com/path",
         "http.request.method" => "GET",
         "http.query" => "foo=bar"
       })
     
       expected: {"http.query"=>"foo=bar", "http.request.method"=>"GET", "http.response.status_code"=>200, "url"=>"http://example.com/path"}
            got: {"http.query"=>{"foo"=>"bar"}, "http.request.method"=>"GET", "http.response.status_code"=>200, "url"=>"http://example.com/path"}
     
       (compared using ==)
     
       Diff:
       @@ -1,4 +1,4 @@
       -"http.query" => "foo=bar",
       +"http.query" => {"foo"=>"bar"},
        "http.request.method" => "GET",
        "http.response.status_code" => 200,
        "url" => "http://example.com/path",
       
     # ./spec/sentry/excon_spec.rb:84:in `block (4 levels) in <top (required)>'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'

  2) Sentry::Excon with tracing enabled with config.send_default_pii = true records breadcrumbs
     Failure/Error: expect(crumb.data[:query]).to eq("foo=bar")
     
       expected: "foo=bar"
            got: {"foo"=>"bar"}
     
       (compared using ==)
     
       Diff:
       @@ -1 +1 @@
       -"foo=bar"
       +"foo" => "bar",
       
     # ./spec/sentry/excon_spec.rb:108:in `block (4 levels) in <top (required)>'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # /usr/local/rvm/gems/default/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'

Finished in 10.32 seconds (files took 1.12 seconds to load)
1069 examples, 2 failures, 1 pending

Failed examples:

rspec ./spec/sentry/excon_spec.rb:66 # Sentry::Excon with tracing enabled with config.send_default_pii = true records the request's span with query string in data
rspec ./spec/sentry/excon_spec.rb:92 # Sentry::Excon with tracing enabled with config.send_default_pii = true records breadcrumbs
```

this is under newer Rubies

**Comment by frederikspang** [comment]
@solnic I'm thinking this may relate to excon 1.0.0 released JUST the other day. 

We've had some specs failing internally as well, post upgrade. 
Like when using webmock, `Host` header had port before, and not after 1.0.0 release.

I'll look into this for the specs!

**Comment by frederikspang** [comment]
@solnic A few seemingly unrelated failures as far as I can tell.
Coverage should be improved to 100% patch as well.

**Comment by frederikspang** [comment]
@solnic All good!

**Pull request merged**